ziti/src/main/kotlin/org/openziti/net/nio/SyncSocketChannel.kt

```diff
@@ -0,0 +1,194 @@
+import org.openziti.net.ZitiSocketFactory
+import org.openziti.util.Logged
+import org.openziti.util.ZitiLog
+import java.io.IOException
+import java.net.*
+import java.nio.ByteBuffer
+import java.nio.channels.AsynchronousSocketChannel
+import java.nio.channels.CompletionHandler
+import java.nio.channels.SocketChannel
+import java.nio.channels.spi.SelectorProvider
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+class SyncSocketChannel(provider: SelectorProvider?) : SocketChannel(provider),
+    Logged by ZitiLog() {
+
+    private lateinit var channelImpl: AsynchronousSocketChannel
+
+    private fun validateConnection() {
+        if (!this::channelImpl.isInitialized || isConnected.not()) IOException("Channel is not connected")
+    }
+
+    override fun implCloseSelectableChannel() {
+        if(this::channelImpl.isInitialized) {
+            channelImpl.close()
+        }
+    }
+
+    override fun implConfigureBlocking(block: Boolean) {
+        TODO("Not yet implemented")
+    }
+
+    override fun read(dst: ByteBuffer?): Int {
+        validateConnection()
+        return channelImpl.read(dst).get()
+    }
+
+    override fun read(dsts: Array<out ByteBuffer>?, offset: Int, length: Int): Long {
+        validateConnection()
+        var response: Long
+
+        runBlocking {
+             response = suspendCoroutine { cont ->
+                 channelImpl.read(
+                     dsts,
+                     offset,
+                     length,
+                     -1, // No read timeout set from here - delegate to underlying channel
+                     TimeUnit.MILLISECONDS,
+                     this,
+                     object : CompletionHandler<Long, CoroutineScope> {
+                         override fun completed(result: Long?, attachment: CoroutineScope?) {
+                             cont.resume(result ?: -1)
+                         }
+
+                         override fun failed(exc: Throwable?, attachment: CoroutineScope?) {
+                             if (null != exc) cont.resumeWithException(exc)
+                             else cont.resumeWithException(IllegalStateException("Channel read failed without an exception"))
+                         }
+                     })
+             }
+        }
+        return response
+    }
+
+    override fun write(src: ByteBuffer?): Int {
+        validateConnection()
+        return channelImpl.write(src).get()
+    }
+
+    override fun write(srcs: Array<out ByteBuffer>?, offset: Int, length: Int): Long {
+        validateConnection()
+        var response: Long
+
+        runBlocking {
+            response = suspendCoroutine { cont ->
+                channelImpl.write(
+                    srcs,
+                    offset,
+                    length,
+                    -1,  // No read timeout set from here - delegate to underlying channel
```

Comment:
AsyncSocketChannel does not handle SO_TIMEOUT option. If app called set SO_TIMEOUT this class need to handle it and call read/write with appropriate timeout

Reply:
I'm not sure about this.  The NIO1 stuff does not support a timeout directly.  From here:  https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/channels/SelectableChannel.html

`A selectable channel is either in blocking mode or in non-blocking mode. In blocking mode, every I/O operation invoked upon the channel will block until it completes. In non-blocking mode an I/O operation will never block and may transfer fewer bytes than were requested or possibly no bytes at all. The blocking mode of a selectable channel may be determined by invoking its isBlocking method.
Newly-created selectable channels are always in blocking mode. Non-blocking mode is most useful in conjunction with selector-based multiplexing. A channel must be placed into non-blocking mode before being registered with a selector, and may not be returned to blocking mode until it has been deregistered.`

The current implementation cannot be changed to non-blocking - it will throw an exception.   I think the behavior here to block until the request is complete is correct.